### PR TITLE
Composable repetition schedules

### LINKED ISSUE #195: Composable repetitions
There are a variety of ad hoc methods on `IO` for repeating actions, either fixed number of times, variable number of times, or for certain durations. We should do something similar as for `Retry`—introduce a `Schedule` abstraction that's very flexible and composable and that can subsume the existing methods and many more.

## Usage API

Fake code to experiment with API:

```scala
// Repeat for n times:
Repeat.repetitions(10)

// Repeat forever & aggregate results
Repeat.collectWith(Nil)(_ :: _)

// Repeat forever every 100 milliseconds:
Repeat.intervals(100.milliseconds)

// Repeat for up to 10 seconds:
Repeat.capped(10.seconds)
```

```scala
val policy =
  (Repeat.repetitions(10) && Repeat.capped(10.seconds) && Repeat.capped(10.seconds))
  *> Repeat.collectWith(Nil)(_ :: _)

val result = IO[E, List[A]] = io.repeat(policy)
```

### PR DESCRIPTION
**Pull request by jdegoes**
Closes #195

### CODE REVIEW

File: core/shared/src/main/scala/scalaz/zio/Repeat.scala [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+    }
+
+  /**
+   * Returns a new schedule that deals with a narrower class of inputs than
+   * this one.
+   */
+  final def contramap[A1 <: A](f: A1 => A): Repeat[A1, B] =
+    new Repeat[A1, B] {
+      type State = self.State
+      val initial = self.initial
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Repeat.scala]: `f andThen self.initial` ?

File: core/shared/src/main/scala/scalaz/zio/Repeat.scala [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    check[B](IO.now(_))(f)
+
+  /**
+   * Returns a new schedule that continues the schedule only until the predicate
+   * is satisfied on the output value of the schedule.
+   */
+  final def untilValue(f: B => Boolean): Repeat[A, B] =
+    check[B](IO.now(_))(!f(_))
+
+  final def combineWith[A1 <: A, C](that: Repeat[A1, C])(g: (Boolean, Boolean) => Boolean,
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Repeat.scala]: Can `g` be dropped? It is not used.

Comment (regiskuckaertz) [reply core/shared/src/main/scala/scalaz/zio/Repeat.scala]: Ha, scratch that, it's `Step.combineWith` that is used in `update`.

File: core/shared/src/main/scala/scalaz/zio/Repeat.scala [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    new Repeat[A1, B] {
+      type State = self.State
+      val initial = self.initial
+      val start   = self.start
+      val value   = self.value
+      val update  = f(self.update)
+    }
+
+  /**
+   * Returns a new repeat schedule with the specified initial value transforemd
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Repeat.scala]: `transformed`

File: core/shared/src/main/scala/scalaz/zio/Repeat.scala [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+  /**
+   * A schedule that repeats forever, returning each input as the output.
+   */
+  final def identity[A]: Repeat[A, A] =
+    Repeat[A, A](IO.now(_), IO.now(Duration.Zero), (a, _) => IO.now(Step.cont(a, Duration.Zero)))
+
+  /**
+   * A schedule that never executes.
+   */
+  final def never: Repeat[Any, Unit] =
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Repeat.scala]: I wonder if it is possible to write a `Repeat[Any, Nothing]`, I guess not without reworking `Step`
```scala
Repeat[Any, Nothing](_ => IO.never, IO.never.const(Duration.Inf), ???)
```

**Comment by jdegoes** [comment]
@regiskuckaertz All right, I addressed most (all) of your feedback and unified `Repeat` with `Retry` since they were so similar. I had to delete the starting delay, since it doesn't make sense for `Retry` (there is no delay before the first failure). In theory it can still be encoded via `initial.delay(start)`, but in practice I never used it for anything, although if anyone uses this to do stuff like `every(TUESDAY)`, they will need this capability.

File: core/shared/src/main/scala/scalaz/zio/Schedule.scala [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+   * A schedule that recurs forever, returning the constant for every output.
+   */
+  final def point[A](a: => A): Schedule[Any, A] = forever.const(a)
+
+  /**
+   * A schedule that never executes. Note that negating this schedule does not
+   * produce a schedule that executes.
+   */
+  final val never: Schedule[Any, Nothing] =
+    Schedule[Nothing, Any, Nothing](IO.never, (_, _) => IO.never)
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Schedule.scala]: 🎉

**Comment by regiskuckaertz** [comment]
That looks really great 👍 and a great learning lesson in API design btw. The only thing I noticed is that there could be a `dimap` but it's just as an FYI.

**Comment by jdegoes** [comment]
@regiskuckaertz Done! Thanks for the review. 🙏 😄

**Pull request merged**